### PR TITLE
Add helper function to unpack constant scalar out of array

### PR DESCRIPTION
**Pull request by robert3005**

### CODE REVIEW

File: vortex-array/src/compute/slice.rs [hunk 1]
```diff
@@ -26,12 +24,8 @@ pub fn slice(array: impl AsRef<ArrayData>, start: usize, stop: usize) -> VortexR
     let array = array.as_ref();
     check_slice_bounds(array, start, stop)?;
 
-    if array
-        .statistics()
-        .get_as(Stat::IsConstant)
-        .unwrap_or_default()
-    {
-        return Ok(ConstantArray::new(scalar_at(array, 0)?, stop - start).into_array());
+    if let Some(rhs) = array.constant() {
```
Comment (a10y) [inline vortex-array/src/compute/slice.rs]:
```suggestion
    if let Some(const_scalar) = array.constant() {
```

`rhs` is an awkward name for this

File: vortex-sampling-compressor/src/compressors/constant.rs [hunk 2]
```diff
@@ -33,7 +32,11 @@ impl EncodingCompressor for ConstantCompressor {
         _ctx: SamplingCompressor<'a>,
     ) -> VortexResult<CompressedArray<'a>> {
         Ok(CompressedArray::compressed(
-            ConstantArray::new(scalar_at(array, 0)?, array.len()).into_array(),
+            ConstantArray::new(
+                array.constant().vortex_expect("checked earlier"),
```
Comment (a10y) [inline vortex-sampling-compressor/src/compressors/constant.rs]:
```suggestion
                array.constant().vortex_expect("ConstantCompressor expects constant array"),
```

File: vortex-array/src/lib.rs [hunk 3]
```diff
@@ -115,6 +115,7 @@ impl ArrayData {
     /// Return scalar value of this array if the array is constant
     pub fn constant(&self) -> Option<Scalar> {
         (self.statistics().get_as::<bool>(Stat::IsConstant)?)
+            // This is safe to unwrap as long as empty arrays aren't constant
```
Comment (a10y) [inline vortex-array/src/lib.rs]: shall we make ConstantArray::new fail if the provided length is 0?

Comment (robert3005) [reply vortex-array/src/lib.rs]: eh... that's a great question... I think not but we shouldn't say it's constant according to statistics, i.e. it's unknown

**Review by gatesn — APPROVED** [review]
Not a fan of the name "constant", but not sure of anything better right now

File: vortex-array/src/lib.rs [hunk 4]
```diff
@@ -108,6 +112,14 @@ impl ArrayData {
         self.with_dyn(|a| a.nbytes())
     }
 
+    /// Return scalar value of this array if the array is constant
+    pub fn constant(&self) -> Option<Scalar> {
```
Comment (lwwmanning) [inline vortex-array/src/lib.rs]: call this `as_constant` instead, I think

Comment (robert3005) [reply vortex-array/src/lib.rs]: conflicted, `constant` is so short

Comment (robert3005) [reply vortex-array/src/lib.rs]: you got it

Comment (lwwmanning) [reply vortex-array/src/lib.rs]: yeah, but we do `as_bool_array`, etc. that returns `Option<&dyn BoolArrayTrait>`. only 3 extra characters, much clearer :D

File: vortex-array/src/stats/statsset.rs [hunk 5]
```diff
@@ -55,9 +55,11 @@ impl StatsSet {
 
     pub fn constant(scalar: Scalar, length: usize) -> Self {
         let mut stats = Self::default();
-        stats.set(Stat::IsConstant, true);
-        stats.set(Stat::IsSorted, true);
-        stats.set(Stat::IsStrictSorted, length <= 1);
+        if length > 0 {
```
Comment (a10y) [inline vortex-array/src/stats/statsset.rs:61]: I think it's weird to flag this just for constant, e.g. empty null array will have all of these set to true

Comment (a10y) [reply vortex-array/src/stats/statsset.rs:61]: isn't it easier to just check that len is non-zero in `as_constant` ?

Comment (robert3005) [reply vortex-array/src/stats/statsset.rs:61]: from purity point of view I think having base cases for empty arrays make sense but at the same time I think we should report stats that are useful. IsContant, IsSorted and IsStrictSorted are not that useful to know for an empty array and then you'd have to do the check. However, you likely will have to do the check anyway

**Review by a10y — APPROVED** [review]

**Pull request merged**